src/core/internal/cppdemangle.d

```diff
@@ -0,0 +1,311 @@
+* See_Also:
+*      https://itanium-cxx-abi.github.io/cxx-abi/abi.html#demangler
+*      https://gcc.gnu.org/onlinedocs/libstdc++/libstdc++-html-USERS-4.3/a01696.html
+*      https://libcxxabi.llvm.org/spec.html
+*      https://docs.microsoft.com/en-us/windows/win32/api/dbghelp/nf-dbghelp-undecoratesymbolname
+*/
+module core.internal.cppdemangle;
+
+/**
+* Demangles C++ mangled names.
+*
+* The output buffer and return will be contains the demangled name of C++ symbol if the demangling is successful.
+*
+* This function will fail and the mangled name in input will be returned if:
+* 1. The symbol to be processed does not follow the Itanium ABI for POSIX or Visual C++ for Windows, or more simply it is not a C++ name
+* 2. The platform is not compatible or the library needed to perform the demangling was not found or linked with the executable
+* 3. The size of the output buffer is not sufficient to contain the demangled name with its prefix
+*
+* This function will fail and null will be returned if:
+* 1. The output buffer is null
+* 2. The function has failed (see list above) and output buffer is unable to contain the mangled name in input
+*
+ * Params:
+ *  mangledName = The string to demangle.
+ *  outputBuffer = The destination buffer, if the size of the destination buffer is <= the size of cppdemangle output the return string would be incomplete.
+ *  prefix = Specifies the prefix to be added to the beginning of the string containing the demangled name
+ *
+ * Returns:
+ *  The demangled name or the original string if the name is not a mangled C++ name.
+ */
+char[] cppdemangle(const(char)[] mangledName, char[] outputBuffer, string prefix = "") @trusted
+{
+    CPPDemangle.initialize();
+    return CPPDemangle.instance.cppdemangle(mangledName, outputBuffer, prefix);
+}
+
+package struct CPPDemangle
+{
+    private __gshared CPPDemangle _instance;
+    private __gshared bool isInitialized;
+
+    version (Posix)
+    {
+        @nogc nothrow extern(C)
+        {
+            private extern(C) char* function(const char* mangledName, char* outputBuffer, size_t* length, int* status) __cxa_demangle;
+        }
+
+        version (OSX)
+            private static immutable names = ["libc++abi.dylib", "libstdc++.dylib"];
+        else
+        {
+            private static immutable names = ["libstdc++.so", "libstdc++.so.6", "libstdc++.so.5",
+                "libc++abi.so", "libc++abi.so.1"];
+        }
+
+        private __gshared void* _handle;
+
+        shared static ~this() {
+            import core.sys.posix.dlfcn : dlclose;
+
+            if (isInitialized)
+            {
+                dlclose(_handle);
+                _handle = null;
+                isInitialized = false;
+            }
+        }
+    }
+
+    version (Windows)
+    {
+        import core.sys.windows.dbghelp : UnDecorateSymbolNameFunc;
+
+        @nogc nothrow extern(System)
+        {
+            private UnDecorateSymbolNameFunc UnDecorateSymbolName;
+        }
```

Comment:
`@nogc nothrow extern(System)` doesn't affect the type of `UnDecorateSymbolName` and `extern(System)` is redundant anyway